x-pack/plugins/observability/public/components/custom_threshold/components/preview_chart/painless_tinymath_parser.ts

```diff
@@ -0,0 +1,205 @@
+
+      // Ignore parentheses if they are opened and closed in the same condition as we put them using ifelse function
+      // Extra parentheses will throw error in Lens
+      // We keep the open parentheses and the closing parentheses that are not part of the same condition
+      if (
+        openParentheses &&
+        openParentheses.length > 0 &&
+        closeParentheses &&
+        closeParentheses.length > 0
+      ) {
+        return result;
+      }
+      // Put back the open parentheses
+      if (openParentheses && openParentheses.length > 0) {
+        result = openParentheses[0] + result;
+      }
+      // Put back the closing parentheses
+      if (closeParentheses && closeParentheses.length > 0) {
+        result = result + closeParentheses[0];
+      }
+      return result;
+    });
+
+    // At the end of the loop, we will have an array of conditions with AND(*) and OR(+)
+    // And to evaluate the whole condition, by wrapping them with ifelse and check if all the conditions are true (>0)
+    const conditions = res.join(' ') + ' > 0';
+    return conditions;
+  }
+
+  getTrueBranch(expression: string) {
+    const lastColonPosition = this.diveIntoString(expression, this.IF, this.ELSE);
+    if (lastColonPosition !== undefined) {
+      return expression.substring(expression.indexOf(this.IF) + 1, lastColonPosition);
+    }
+    return '';
+  }
+
+  getFalseBrach(expression: string) {
+    const lastColonPosition = this.diveIntoString(expression, this.IF, this.ELSE);
+    if (lastColonPosition !== undefined) {
+      return expression.substring(lastColonPosition + 1);
+    }
+    return '';
+  }
+
+  getCondition(expression: string): string {
+    return expression.substring(0, expression.indexOf(this.IF));
+  }
+
+  parse(): string {
+    const recursiveParseConditions = (expression: string): string => {
+      const condition = this.parseCondition(this.getCondition(expression));
+      const trueBranch = this.getTrueBranch(expression);
+      const falseBranch = this.getFalseBrach(expression);
+
+      let fullCondition = `ifelse(${condition}, ${trueBranch}, ${falseBranch})`;
+
+      if (trueBranch.includes(this.IF)) {
+        const parsedTrueBranch = recursiveParseConditions(trueBranch);
+        fullCondition = fullCondition.replace(trueBranch, parsedTrueBranch);
+      }
+      if (falseBranch.includes(this.IF)) {
+        const parsedFalseBranch = recursiveParseConditions(falseBranch);
+        fullCondition = fullCondition.replace(falseBranch, parsedFalseBranch);
+      }
+      const result = this.replaceCharactersWithAggMap(fullCondition, this.aggMap);
+      return result;
+    };
+
+    if (!this.equation || this.equation === '') throw new Error('Invalid Equation');
+
+    if (this.equation.includes(this.IF)) {
+      return recursiveParseConditions(this.equation);
+    }
+    return this.replaceCharactersWithAggMap(this.equation, this.aggMap);
```

Comment:
In relation to above [comment](https://github.com/elastic/kibana/pull/169007/files#r1390905851), since it's only replacing characters with actual aggregations, I think `replaceCharactersWithAggMap` makes sense to reflect what the function is doing. We are not parsing the equation in this function.